File: internal/db/models/network/switch.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package networkModels

import (
	"time"
)

type ManualSwitch struct {
	ID     uint   `json:"id" gorm:"primaryKey;autoIncrement"`
	Name   string `json:"name" gorm:"unique;not null"`
	Bridge string `json:"bridge" gorm:"unique;not null"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

type StandardSwitch struct {
	ID         uint   `json:"id" gorm:"primaryKey;autoIncrement"`
	Name       string `json:"name" gorm:"unique;not null"`
	BridgeName string `json:"bridgeName" gorm:"unique;not null"`
	MTU        int    `json:"mtu" gorm:"default:1500"`
	VLAN       int    `json:"vlan" gorm:"default:0"`
	Address    string `json:"address"`
	Address6   string `json:"address6"`

	AddressID  *uint   `json:"addressId" gorm:"column:address_object_id"`
	AddressObj *Object `json:"addressObj" gorm:"foreignKey:AddressID"`

	Address6ID  *uint   `json:"address6Id" gorm:"column:address6_object_id"`
	Address6Obj *Object `json:"address6Obj" gorm:"foreignKey:Address6ID"`

	NetworkID  *uint   `json:"networkId" gorm:"column:network_object_id"`
	NetworkObj *Object `json:"networkObj" gorm:"foreignKey:NetworkID"`

	Network6ID  *uint   `json:"network6Id" gorm:"column:network6_object_id"`
	Network6Obj *Object `json:"network6Obj" gorm:"foreignKey:Network6ID"`

	GatewayAddressID  *uint   `json:"gatewayAddressId" gorm:"column:gateway_address_object_id"`
	GatewayAddressObj *Object `json:"gatewayAddressObj" gorm:"foreignKey:GatewayAddressID"`

	Gateway6AddressID  *uint   `json:"gateway6AddressId" gorm:"column:gateway6_address_object_id"`
	Gateway6AddressObj *Object `json:"gateway6AddressObj" gorm:"foreignKey:Gateway6AddressID"`

	NetworkManual  string `json:"networkManual" gorm:"column:network_manual"`
	Network6Manual string `json:"network6Manual" gorm:"column:network6_manual"`
	GatewayManual  string `json:"gatewayManual" gorm:"column:gateway_manual"`
	Gateway6Manual string `json:"gateway6Manual" gorm:"column:gateway6_manual"`

	DisableIPv6  bool `json:"disableIPv6" gorm:"default:false"`
	Private      bool `json:"private" gorm:"default:false"`
	DefaultRoute bool `json:"defaultRoute" gorm:"default:false"`

	Ports []NetworkPort `json:"ports" gorm:"foreignKey:SwitchID;constraint:OnDelete:CASCADE"`

	DHCP  bool `json:"dhcp" gorm:"default:false"`
	SLAAC bool `json:"slaac" gorm:"default:false"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

type NetworkPort struct {
	ID       int            `json:"id" gorm:"primaryKey;autoIncrement"`
	Name     string         `json:"name" gorm:"not null"`
	SwitchID uint           `json:"switchId" gorm:"not null"`
	Switch   StandardSwitch `gorm:"foreignKey:SwitchID"`
}

// StandardSwitchManualAddresses carries raw, manually-typed address values for a
// standard switch as an alternative to selecting NetworkObjects. Each field is
// mutually exclusive with its corresponding object ID at the service layer.
type StandardSwitchManualAddresses struct {
	Network4 string
	Gateway4 string
	Network6 string
	Gateway6 string
}

func (sw *StandardSwitch) Network(v int) string {
	if v == 4 {
		if sw.NetworkObj != nil && len(sw.NetworkObj.Entries) > 0 {
			return sw.NetworkObj.Entries[0].Value
		}
		return sw.NetworkManual
	}

	if v == 6 {
		if sw.Network6Obj != nil && len(sw.Network6Obj.Entries) > 0 {
			return sw.Network6Obj.Entries[0].Value
		}
		return sw.Network6Manual
	}

	return ""
}

func (sw *StandardSwitch) Gateway(v int) string {
	if v == 4 {
		if sw.GatewayAddressObj != nil && len(sw.GatewayAddressObj.Entries) > 0 {
			return sw.GatewayAddressObj.Entries[0].Value
		}
		return sw.GatewayManual
	}

	if v == 6 {
		if sw.Gateway6AddressObj != nil && len(sw.Gateway6AddressObj.Entries) > 0 {
			return sw.Gateway6AddressObj.Entries[0].Value
		}
		return sw.Gateway6Manual
	}

	return ""
}

func (sw *StandardSwitch) IPv4() string {
	if sw.AddressObj != nil && len(sw.AddressObj.Entries) > 0 {
		return sw.AddressObj.Entries[0].Value
	}
	return ""
}

func (sw *StandardSwitch) IPv6() string {
	if sw.Address6Obj != nil && len(sw.Address6Obj.Entries) > 0 {
		return sw.Address6Obj.Entries[0].Value
	}
	return ""
}

File: web/src/lib/types/network/switch.ts
/**
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2025 The FreeBSD Foundation.
 *
 * This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
 * of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
 * under sponsorship from the FreeBSD Foundation.
 */

import { z } from 'zod/v4';
import { NetworkObjectSchema, type NetworkObject } from './object';
import type { Row } from '../components/tree-table';

const nullableString = z
	.string()
	.nullish()
	.transform((value) => value ?? '');

export const VLANPortPolicySchema = z.object({
	mode: z.enum(['', 'access', 'trunk']),
	untaggedVlan: z.number().int().min(1).max(4094).optional(),
	taggedVlans: z.array(z.number().int().min(1).max(4094)).default([])
});

export const NetworkPortSchema = z.object({
	id: z.number().int().positive(),
	name: z.string(),
	switchId: z.number().int().positive(),
	vlanPolicy: VLANPortPolicySchema
});

export const StandardSwitchRCConflictSchema = z.object({
	code: z.enum([
		'standard_switch_member_rc_l3_conflict',
		'standard_switch_member_bridge_ownership_conflict',
		'standard_switch_member_rc_inspection_unavailable',
		'standard_switch_runtime_bridge_inspection_unavailable'
	]),
	port: z.string().optional(),
	member: z.string().optional(),
	conflictingBridge: z.string().optional(),
	dhcp: z.boolean().optional(),
	slaac: z.boolean().optional(),
	staticIPv4: z.boolean().optional(),
	staticIPv6: z.boolean().optional(),
	aliasesIPv4: z.boolean().optional(),
	aliasesIPv6: z.boolean().optional(),
	rcConfigured: z.boolean().optional(),
	runtimeAttached: z.boolean().optional()
});

export const StandardSwitchRCConflictsSchema = z.array(StandardSwitchRCConflictSchema);

export const StandardSwitchSchema = z.object({
	id: z.number().int().positive(),
	name: z.string(),
	bridgeName: z.string(),
	mtu: z.number().int(),
	vlan: z.number().int(),
	private: z.boolean(),
	address: z.string(),
	address6: z.string(),
	addressObj: NetworkObjectSchema.nullable(),
	address6Obj: NetworkObjectSchema.nullable(),
	networkObj: NetworkObjectSchema.nullable(),
	network6Obj: NetworkObjectSchema.nullable(),
	gatewayAddressObj: NetworkObjectSchema.nullable(),
	gateway6AddressObj: NetworkObjectSchema.nullable(),
	networkManual: nullableString,
	network6Manual: nullableString,
	gatewayManual: nullableString,
	gateway6Manual: nullableString,
	ports: z.array(NetworkPortSchema),
	bridgeMacMode: z.enum(['port', 'object']),
	bridgeMacSourcePort: nullableString,
	bridgeMacObjectId: z.number().int().positive().nullable(),
	bridgeMacObject: NetworkObjectSchema.nullable(),
	dhcp: z.boolean(),
	slaac: z.boolean(),
	disableIPv6: z.boolean(),
	defaultRoute: z.boolean(),
	defaultRoute6: z.boolean().default(false),
	disableBridgeOffloads: z.boolean(),
	vlanFiltering: z.boolean().default(false),
	defaultAccessVlan: z.number().int().min(1).max(4094).nullable().default(null),
	hostVlan: z.number().int().min(1).max(4094).nullable().default(null)
});

export const ManualSwitchSchema = z.object({
	id: z.number().int().positive(),
	name: z.string(),
	bridge: z.string(),
	vlanFiltering: z.boolean().default(false),
	defaultAccessVlan: z.number().int().min(1).max(4094).nullable().default(null),
	vlanStateAvailable: z.boolean(),
	vlanStateError: z.string().default(''),
	createdAt: z.string(),
	updatedAt: z.string()
});

export const SwitchListSchema = z.object({
	standard: z.array(StandardSwitchSchema),
	manual: z.array(ManualSwitchSchema)
});

export type StandardSwitchRCConflict = z.infer<typeof StandardSwitchRCConflictSchema>;
export type VLANPortPolicy = z.infer<typeof VLANPortPolicySchema>;
export type StandardSwitch = z.infer<typeof StandardSwitchSchema>;
export type ManualSwitch = z.infer<typeof ManualSwitchSchema>;
export type NetworkSwitch = StandardSwitch | ManualSwitch;
export type SwitchList = z.infer<typeof SwitchListSchema>;

export function isSwitchRuntimeAvailable(networkSwitch: NetworkSwitch): boolean {
	return !('vlanStateAvailable' in networkSwitch) || networkSwitch.vlanStateAvailable;
}

export function isSwitchVMCompatible(networkSwitch: NetworkSwitch): boolean {
	return (
		isSwitchRuntimeAvailable(networkSwitch) &&
		(!networkSwitch.vlanFiltering || networkSwitch.defaultAccessVlan !== null)
	);
}

export interface SwitchRow extends Row {
	id: number;
	name: string;
	mtu: number;
	vlan: number | '-';
	ports: Array<{ name: string; vlanPolicy: VLANPortPolicy }>;
	portsOnly: string[];
	bridgeMacMode: 'port' | 'object';
	bridgeMacSourcePort: string;
	bridgeMacObjectId: number | null;
	bridgeMacObject?: NetworkObject | null;
	networkObj?: NetworkObject;
	networkManual?: string;
	network6Obj?: NetworkObject;
	network6Manual?: string;
	gatewayAddressObj?: NetworkObject;
	gatewayManual?: string;
	gateway6AddressObj?: NetworkObject;
	gateway6Manual?: string;
	disableIPv6: boolean;
	private: boolean;
	dhcp: boolean;
	slaac: boolean;
	defaultRoute: boolean;
	defaultRoute6: boolean;
	disableBridgeOffloads: boolean;
	vlanFiltering: boolean;
	defaultAccessVlan: number | null;
	hostVlan: number | null;
	children?: SwitchRow[];
}

export interface ManualSwitchRow extends Row {
	id: number;
	name: string;
	bridge: string;
	vlanFiltering: boolean;
	defaultAccessVlan: number | null;
	vlanStateAvailable: boolean;
	vlanStateError: string;
	children?: ManualSwitchRow[];
}

export function emptySwitchList(): SwitchList {
	return { standard: [], manual: [] };
}

export function isSwitchList(value: unknown): value is SwitchList {
	return SwitchListSchema.safeParse(value).success;
}
